vm/stardust/vm_stardust_test.go

```diff
@@ -1264,6 +1264,1512 @@ func TestStardustTransactionExecution(t *testing.T) {
+				unlocksFunc: func(sigs []iotago.Signature, testAddresses []iotago.Address) iotago.Unlocks {
+					return iotago.Unlocks{
+						&iotago.MultiUnlock{
+							Unlocks: []iotago.Unlock{
+								&iotago.SignatureUnlock{Signature: sigs[0]},
+								&iotago.SignatureUnlock{Signature: sigs[1]},
+							},
+						},
+						&iotago.MultiUnlock{
+							Unlocks: []iotago.Unlock{
+								&iotago.EmptyUnlock{},
+								&iotago.SignatureUnlock{Signature: sigs[2]},
+								&iotago.SignatureUnlock{Signature: sigs[3]},
+							},
+						},
+					}
+				},
+				wantEncodeErr:  nil,
+				wantExecuteErr: nil,
+			}
+		}(),
+
+		// ok - multiple MultiAddresses in one TX - signature reuse in different multi unlocks
+		func() test {
+			return test{
+				name:           "ok - multiple MultiAddresses in one TX - signature reuse in different multi unlocks",
+				ed25519AddrCnt: 4,
+				addressesFunc: func(ed25519Addresses []iotago.Address) []iotago.Address {
+					return []iotago.Address{
+						&iotago.MultiAddress{
+							Addresses: []*iotago.AddressWithWeight{
+								{
+									Address: ed25519Addresses[0],
+									Weight:  1,
+								},
+								{
+									Address: ed25519Addresses[1],
+									Weight:  1,
+								},
+							},
+							Threshold: 2,
+						},
+						&iotago.MultiAddress{
+							Addresses: []*iotago.AddressWithWeight{
+								{
+									// optional
+									Address: ed25519Addresses[0],
+									Weight:  1,
+								},
+								{
+									// optional
+									Address: ed25519Addresses[2],
+									Weight:  1,
+								},
+								{
+									// mandatory
+									Address: ed25519Addresses[3],
+									Weight:  2,
+								},
+							},
+							Threshold: 3,
+						},
+					}
+				},
+				inputsFunc: func(ed25519Addresses []iotago.Address, testAddresses []iotago.Address) []iotago.Output {
+					return []iotago.Output{
+						&iotago.BasicOutput{
+							Amount: defaultAmount,
+							Conditions: iotago.BasicOutputUnlockConditions{
+								&iotago.AddressUnlockCondition{Address: testAddresses[0]},
+							},
+						},
+						&iotago.BasicOutput{
+							Amount: defaultAmount,
+							Conditions: iotago.BasicOutputUnlockConditions{
+								&iotago.AddressUnlockCondition{Address: testAddresses[1]},
+							},
+						},
+					}
+				},
+				outputsFunc: nil,
+				unlocksFunc: func(sigs []iotago.Signature, testAddresses []iotago.Address) iotago.Unlocks {
+					return iotago.Unlocks{
+						&iotago.MultiUnlock{
+							Unlocks: []iotago.Unlock{
+								&iotago.SignatureUnlock{Signature: sigs[0]},
+								&iotago.SignatureUnlock{Signature: sigs[1]},
+							},
+						},
+						&iotago.MultiUnlock{
+							Unlocks: []iotago.Unlock{
+								&iotago.SignatureUnlock{Signature: sigs[0]},
+								&iotago.EmptyUnlock{},
+								&iotago.SignatureUnlock{Signature: sigs[3]},
+							},
+						},
+					}
+				},
+				wantEncodeErr:  nil,
+				wantExecuteErr: nil,
+			}
+		}(),
+
+		//
+		// Syntactical checks tests
```

Comment:
There are a bunch of syntactical tests in here that, if I'm not mistaken, could be moved to a simpler serialize/deserialize test, because some of the test stuff for other tests is not required for the syntactical checks. That would already clean up this test function a bit and make the whole test easier to understand. The whole thing is not exactly trivial, so keeping it somewhat minimal would be wonderful for future maintenance.

Reply:
I will fix this in another PR :thumbsup: